### PR TITLE
Consolidate FindControl methods and add tests

### PR DESCRIPTION
**Pull request by RussKie**
Consolidate FindControl methods and add tests

Replaces 947ba4330b79e1c8de576abc67957b54cda7d7bd

### CODE REVIEW

**Comment by RussKie** [comment]
I've run it in isolation (except for the gource, which I don't have)

**Comment by RussKie** [comment]
@drewnoakes would it be ok if we merge this?

**Comment by drewnoakes** [comment]
Up to you. It's a cherry pick from the many commits in #5087. I can rebase it away and deal with any conflicts.

What do you need in order to merge #5087? Would you like me to create issues for outstanding tasks for 3.0 that follow on from that PR? That might you estimate the 'debt' of merging and whether we can pay it off, plus time for adequate testing, before 3.0 RC1. I have a list of issues in my notes and can write them up.

**Comment by RussKie** [comment]
Let's merge your PR :)

**Comment by drewnoakes** [comment]
I'll go ahead and document the issues. Let me do that before we merge. If there's anything major I'll fix it up in the next few days.

**Comment by drewnoakes** [comment]
@RussKie, I fixed up a few outstanding issues last night and documented the three remaining bugs that must be fixed for 3.0 if #5087 is merged:

- Browsing submodule repo can introduce UI pauses #5166
- Graph column width sometimes appears too narrow #5167
- Graph column sometimes not painting last row #5168 

I have been reviewing the flow of data between threads that collaborate to load, process and render revisions and that work should address all three of these issues in one shot.

I will be available to address any other bugs that may arise once this is exposed to wider testing.

**Comment by RussKie** [comment]
Thank you. Let's track and work on those separately.

File: GitUI/GitUIExtensions.cs [hunk 1]
```diff
@@ -198,6 +198,49 @@ private static IEnumerable<TreeNode> AllNodes(this TreeNodeCollection nodes)
...
+        {
+            return FindDescendants(control).OfType<T>();
+        }
+
+        /// <summary>
+        /// Finds the first descendent of <paramref name="control"/> that has type
+        /// <typeparamref name="T"/> and satisfies <paramref name="predicate"/>.
+        /// </summary>
+        [CanBeNull]
+        public static T FindDescendantsOfType<T>(this Control control, Func<T, bool> predicate) where T : Control
```
Comment (sharwell) [inline GitUI/GitUIExtensions.cs]: ❗️ This should be singular.

:bulb: In addition making the method singular, consider using a name that indicates `default(T)` could be returned (e.g. `*OrDefault` or `Try*`). Alternately, you could just remove the use of `FirstOrDefault` and consolidate this with the previous method (making `predicate` optional and act as `x=>true` when omitted).

Comment (RussKie) [reply GitUI/GitUIExtensions.cs]: Thanks Sam.
Holding off until Drew merges his PR in, then will address your comments.

Comment (drewnoakes) [reply GitUI/GitUIExtensions.cs]: The pluralisation of that method name was fixed in #5087.

**Comment by drewnoakes** [comment]
This change is now on master.

**Pull request closed without merging**